Try/add drag handle to select mode

draganescu opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Advances #24092

Adds a drag handle to the block selector button when in navigation mode. 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots


https://user-images.githubusercontent.com/107534/110501772-2e80a580-8103-11eb-81d5-c9aa3a43c8d7.mp4


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

jasmussen commented:
Whoah. A Sunday PR, and an impressive one at that! This is what I see:

![select mode](https://user-images.githubusercontent.com/1204802/107149499-3b21aa80-6959-11eb-86c9-3cf263521756.gif)

A few things come to mind, that don't necessarily have to be part of this PR, but are inspired by how good this feels:

1. We should consider adding the block icon before the drag handle in select mode, so that when you lift the block off the page, the black chip that you then see matches in structure.
2. Should the black chip _be_ the "select mode label"? I.e. when in select mode, the block toolbar is the same visual style as the floating chip?

Those are small nitpicks. But one thing I'd love to see, is that the entire _block itself_ is the draggable area, in addition to the drag handle.

![selecting](https://user-images.githubusercontent.com/1204802/107149604-c8fd9580-6959-11eb-8a97-15bf66d626ef.gif)

Right now when you click and drag on the block itself, nothing happens. So it seems like it could be the draggable handle instead. The primary tricky thing here is that the 2nd _click_ on the selected block (i.e. click a selected block twice) should still go back to edit mode. So potentially we need to differentiate between ondrag and onclick, and only let the former actually drag.

Nice work! 👌

talldan commented:
The branch seems to have all the commits from the `try/menus-cleanup` branch. I guess it wasn't based off of master?

draganescu commented:
>  I guess it wasn't based off of master?
Yes @talldan I was testing the new nav screen design PR and branched from it. I'll fix it but I wanted it to be visible b/c it really feels nice and I was curious if I'm biased :)

Commit:
Adds drag handle to block selection button in navigation mode

mtias commented:
@jasmussen @pablohoneyhoney would it make sense to use the dark background for the selection state the moment you press "Escape"?

jasmussen commented:
> would it make sense to use the dark background for the selection state the moment you press "Escape"?

Yes, that would be nice. I can probably push code directly to this branch to that effect?

jasmussen committed:
Make handle dark.

jasmussen commented:
I made the handle dark, I think it works well:

![dark handle](https://user-images.githubusercontent.com/1204802/109142408-23bb2d80-775f-11eb-9553-01e67991edb3.gif)

But now that it's dark, should it be identical to the one you see when you're dragging? I.e. in this case, just show a paragraph icon and a drag handle? (I think yes)

We could also increase the drop shadow _while dragging_, to better ambiguate between resting and lifting. What do you think Andrei?

mtias commented:
Nice!

> But now that it's dark, should it be identical to the one you see when you're dragging? I.e. in this case, just show a paragraph icon and a drag handle? (I think yes)

Yes, it'd make sense to try. Could also make sense to have [icon] [drag] Label. With label disappearing when you start dragging but the icon and drag handle remaining exactly the same.

jasmussen commented:
> Yes, it'd make sense to try. Could also make sense to have [icon] [drag] Label. With label disappearing when you start dragging but the icon and drag handle remaining exactly the same.

Yes, that's probably best 👍 👍

draganescu commented:
Lovely additions @jasmussen

jasmussen committed:
Dropper shadow.

jasmussen commented:
Made the shadow droppier, and fixed a color issue with the grabby handle active state:

![drag shadow](https://user-images.githubusercontent.com/1204802/109279266-d4860300-7819-11eb-8c05-7e59b288362b.gif)

draganescu commented:
I am thinking of stealing from #23497 to see how that works in this slect mode context.

draganescu commented:
My opinion is to merge this like it is and follow up with a secondary PR to explore long press drag for the blocks. Should we change the block popover (the handle in navigation mode) to look exactly like the 1st part of the block toolbar? Or is the block name better in this mode than an icon?

Commit:
makes the block selection button look like the daggable chip plus block name

draganescu commented:
@jasmussen I went for (1) [block-type-icon] [drag-handle] Label and also implemented the block selection button to be like the drag chip based on flex, bolted some CSS to make the transition when dragging look normal, and here it is: I updated the screen recording in the PRs description.

What do you think?

draganescu commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -614,28 +614,58 @@
+	padding: 0 20px;
 	z-index: z-index(".block-editor-block-list__block-selection-button");
+	// Dark block UI appearance.
+	border-radius: $radius-block-ui;
```
The comment seems to be in the wrong place.

draganescu commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -614,28 +614,58 @@
+	.components-button.has-icon.block-selection-button_drag-handle {
+		cursor: grab;
+		padding: 0;
+		height: 24px;
```
I wonder why did I need to bolt these magic numbers here.

draganescu commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -614,28 +614,58 @@
 .block-editor-block-list__block-selection-button {
-	display: block;
+	display: inline-flex;
+	padding: 0 20px;
```
This was kinda too hard to get right, none of the `$grid-unit` options seemed to fit right "visually" with the chip once the drag started.

Commit:
fix slight off vertical centering

jasmussen commented:
This is what I see:

<img width="420" alt="Screenshot 2021-03-09 at 18 31 04" src="https://user-images.githubusercontent.com/1204802/110512713-bebfea00-8105-11eb-91d1-8181b54ebd59.png">

Great. I think I need to push some tweaks to get some of the margin stuff right, it feels slightly off even if it isn't necessarily. I'll massage those pixels tomorrow!

jasmussen committed:
Make select mode chip accurate.

jasmussen committed:
Make draggable chip accurate.

jasmussen commented:
Alright, pushed a few things to get the metrics right:

<img width="1234" alt="Screenshot 2021-03-10 at 08 06 51" src="https://user-images.githubusercontent.com/1204802/110591725-b5706500-8179-11eb-9909-c2fc9a0f7c23.png">

<img width="1319" alt="Screenshot 2021-03-10 at 08 19 38" src="https://user-images.githubusercontent.com/1204802/110591730-b73a2880-8179-11eb-9d0d-bb2904555d2f.png">

![drag](https://user-images.githubusercontent.com/1204802/110591738-b86b5580-8179-11eb-9f35-0ceb9b6f8ee7.gif)

I think this is good to land.

gwwar reviewed (approved):
This feels pretty nice and code looks reasonable. Was there anything in particular to focus on here for testing @draganescu ?

For other reviewers 👀 , a quick way of getting into select mode is by pressing <kbd>esc</kbd> while a block is focused.

https://user-images.githubusercontent.com/1270189/110854920-81c13800-826a-11eb-9b3a-c7100e664147.mp4

jasmussen commented:
🔥

mtias commented:
Nice!
